### PR TITLE
feat(container): add self registration of container in constructor, t…

### PR DESCRIPTION
**Pull request by Tolowe**
…est, and document it

### CODE REVIEW

**Comment by thohoh** [comment]
Thank your for the PR, I'll take a look tomorrow. 

You can try implementing this [one](https://github.com/typesoft/container-ioc/issues/102) great way to learn important stuff in Typescript.

File: src/lib/container.interface.ts [hunk 1]
```diff
@@ -6,6 +6,8 @@ export interface IContainerOptions {
 }
 
 export interface IContainer {
+    [key: string]: any;
```
Comment (thohoh) [inline src/lib/container.interface.ts]: Doing that it's a bad practise. Interface (aka super type) should define only public methods. Unless you use interface for defining structure of value objects. Please remove that.

File: src/lib/container.ts [hunk 2]
```diff
@@ -21,6 +21,7 @@ export class Container implements IContainer {
             this.parent = <IContainer> options.parent;
             this.defaultLifeTime = options.defaultLifeTime || this.defaultLifeTime;
         }
+        this.register({ token: TContainer, useValue: this });
```
Comment (thohoh) [inline src/lib/container.ts]: You did right! But I'm gonna release that simplified Reflect API which is gonna allow us to inject stuff by just specifying it's type like this `constructor (private container: Container) {}`. Please use Container as the token in this case.

File: src/tests/container.spec.ts [hunk 3]
```diff
@@ -449,6 +464,14 @@ describe('Container', () => {
 
             expect(instance1).not.to.be.equal(instance2);
         });
+
+        it('should register itself for injection', () => {
+            // arrange, act
+            const actual = container.resolve(TContainer);
+            // assert
```
Comment (thohoh) [inline src/tests/container.spec.ts]: You can't deny that comments like this one are obvious. Clearly named variables and methods won't need any comments above them. Please remove them.

**Comment by thohoh** [comment]
@Tolowe Tt's better to just close this PR, then pull changes and then fix yours and open new PR.

**Comment by Tolowe** [comment]
@thohoh I think I have everything fixed as you mentioned.

**Comment by thohoh** [comment]
@Tolowe Thank you!

**Pull request merged**